### PR TITLE
[ART-2617] use upstream-equivalent FROM images when a flag is set in group.yaml

### PR DESCRIPTION
**Pull request by locriandev**
Ref. https://issues.redhat.com/browse/ART-2617

### CODE REVIEW

File: doozerlib/distgit.py [hunk 1]
```diff
@@ -1612,6 +1612,50 @@ def _mapped_image_for_assembly_build(self, parent_images, i):
         unique_pullspec += f':{parent_build_nvr["version"]}-{parent_build_nvr["release"]}'
         return unique_pullspec
 
+    def _mapped_image_from_stream(self, image, original_parent, dfp):
```
Comment (sosiouxme) [inline doozerlib/distgit.py:1615]: could there be some `logger.debug` statements in here, and probably even a `logger.info` for when the upstream does override our config? in my tests it wasn't easy to see why the upstream override isn't happening as expected. actually i don't think it's even getting here

File: doozerlib/distgit.py [hunk 2]
```diff
@@ -1612,6 +1612,50 @@ def _mapped_image_for_assembly_build(self, parent_images, i):
...
+                dfp.add_lines(
+                    "",
+                    "# Parent images were rebased matching upstream equivalent that didn't match ART's config",
+                    "",
+                    at_start=True)
+            return mapped_image
+
+        except ChildProcessError:
+            # It doesn't. Emit a warning and do typical stream resolution
+            logger.warning(f'Could not match upstream parent {original_parent}')
```
Comment (sosiouxme) [inline doozerlib/distgit.py:1649]: `self.logger`

File: doozerlib/distgit.py [hunk 3]
```diff
@@ -1612,6 +1612,50 @@ def _mapped_image_for_assembly_build(self, parent_images, i):
...
+            stream = self.runtime.resolve_stream(image.stream)
```
Comment (sosiouxme) [inline doozerlib/distgit.py:1650]: already did this...?

**Review by sosiouxme — CHANGES_REQUESTED** [review]
couldn't get it to work as expected, will probably need a better review later.

when you are refactoring into smaller methods, that's a time i'd really like to see some unit tests added for them (which this file is sorely lacking, due in part to its giant methods).

File: doozerlib/distgit.py [hunk 4]
```diff
@@ -1506,10 +1509,200 @@ def _clean_repos(self, dfp):
...
+
+    def _mapped_image_from_stream(self, image, original_parent, dfp):
+        stream = self.runtime.resolve_stream(image.stream)
+
+        if not self.runtime.group_config.canonical_builders_from_upstream:
+            # Do typical stream resolution.
+            return stream.image
+
+        # When canonical_builders_from_upstream flag is set, try to match upstream FROM
+        original_parent = original_parent
```
[the remark this answers is not in the document]
Comment (locriandev) [reply doozerlib/distgit.py:1623]: Removed.

File: doozerlib/distgit.py [hunk 5]
```diff
@@ -1506,10 +1509,200 @@ def _clean_repos(self, dfp):
...
+    def _mapped_image_for_assembly_build(self, parent_images, i):
+        # When rebasing for an assembly build, we want to use the same parent image
+        # as our corresponding basis image. To that end, we cannot rely on a stream.yml
+        # entry -- which usually refers to a floating tag. Instead, we look up the latest
+        # build of this image, relative to the assembly basis event, in brew. It will have
+        # information on the exact parent images used at the time. We want to use that
+        # specific sha.
+        # If you are here trying to figure out how to change this behavior, you should
+        # consider using 'from!:' in the assembly metadata for this component. This will
+        # all you to fully pin the parent images (e.g. {'from!:' ['image': <pullspec>] })
```
[the remark this answers is not in the document]
Comment (locriandev) [reply doozerlib/distgit.py:1563]: Fixed

File: doozerlib/distgit.py [hunk 6]
```diff
@@ -1506,10 +1509,200 @@ def _clean_repos(self, dfp):
...
+            return stream.image
+
+        # When canonical_builders_from_upstream flag is set, try to match upstream FROM
+        original_parent = original_parent
+        cmd = f'oc image info {original_parent} -o json'
+        out, _ = exectools.cmd_assert(cmd, retries=3)
+        labels = json.loads(out)['config']['config']['Labels']
+        builder_tag = f'{labels["version"]}-{labels["release"]}'
+
+        # Verify wether the image exists
```
[the remark this answers is not in the document]
Comment (locriandev) [reply doozerlib/distgit.py:1629]: Fixed

File: doozerlib/distgit.py (hunk 2 above)
Comment (locriandev) [reply doozerlib/distgit.py:1649]: Fixed

File: doozerlib/distgit.py (hunk 3 above)
Comment (locriandev) [reply doozerlib/distgit.py:1650]: Right. Removed

File: doozerlib/distgit.py [hunk 7]
```diff
@@ -1506,10 +1509,200 @@ def _clean_repos(self, dfp):
...
+        upstream_equivalent = f'registry-proxy.engineering.redhat.com/rh-osbs/' \
+                              f'openshift-golang-builder:{builder_tag}'
+        cmd = f'oc image info {upstream_equivalent} --filter-by-os linux/amd64 -o json'
```
[the remark this answers is not in the document]
Comment (locriandev) [reply doozerlib/distgit.py:1650]: Yes, this code is executed only for images specified with `builder: stream` (and for assembly `stream` builds...)

[the remark this answers is not in the document]
Comment (locriandev) [reply doozerlib/distgit.py]: Fixed in https://github.com/openshift/doozer/pull/646/commits/ac3e68347c5858224b352272133b4bbb4273ad4d

File: doozerlib/distgit.py (hunk 1 above)
Comment (locriandev) [reply doozerlib/distgit.py:1634]: Added in https://github.com/openshift/doozer/pull/646/commits/98d9760e0b1b4eb25f02588e22a6a1e6b4fd8b3b

**Review by sosiouxme — APPROVED** [review]

**Pull request merged**